### PR TITLE
Add Action<T> overloads to methods that takes a Closure on Project

### PR DESCRIPTION
**Pull request by eskatos**
For the sake of type-safety and better Kotlin support.

See https://github.com/gradle/gradle-script-kotlin/issues/122

### CODE REVIEW

**Comment by bamboo** [comment]
Looks good. We have to wait for the `@HasImplicitReceiver` annotation to take effect before merging it as the `dependencies` and `repositories` overloads will probably break the included version of gradle-script-kotlin.

**Comment by bamboo** [comment]
An alternative course of action to mitigate risk and collect the reward earlier would be to split the PR to keep `dependencies` and `repositories` for later.

**Comment by eskatos** [comment]
@bamboo removed the `repositories()` and `dependencies()` overload.
Could you have another look?

File: subprojects/core/src/main/java/org/gradle/api/Project.java [hunk 1]
```diff
@@ -784,6 +802,31 @@
     ConfigurableFileTree fileTree(Object baseDir, Closure configureClosure);
 
     /**
+     * <p>Creates a new {@code ConfigurableFileTree} using the given base directory. The given baseDir path is evaluated
+     * as per {@link #file(Object)}. The action will be used to configure the new file tree. Example:</p>
+     *
+     * <pre autoTested=''>
+     * def myTree = fileTree('src') {
+     *    exclude '**&#47;.data/**'
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/api/Project.java]: Although the syntax is correct and that's what we expect build authors using Groovy to write, this example actually represents a call to the `Closure` version of this method.

I think that's ok for now but we should consider reviewing the javadoc samples on `Action<T>` overloads in the future and use Kotlin syntax on those.

**Pull request merged**